### ISSUE TITLE
The pwsh CLI should support the -l / --login option for macOS compatibility, e.g. in VSCode

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#7704]
Related: #3743

On macOS, unlike on Linux, all (user-)default shell instances are started as _login_ shells.

There are two mechanisms for invoking a shell as a login shell: (a) setting its `argv[0]` to contain `-` as the 1st character (that's what Terminal.app does) or (b) passing the `-l` (or `--login`) option to the shell executable.

If you make `pwsh` your _default shell_ on macOS, certain applications may invoke `pwsh` with `-l`, which currently _breaks_ the invocation.

Specifically, Visual Studio Code does that, so on a macOS system that uses `/usr/bin/pwsh` as the default shell, _the integrated terminal fails to load_ (unless you happen to have the PowerShell extension installed and happen to open a `*.ps1` file).

It would be sufficient for `pwsh` to _accept_  `-l` / `--login` and otherwise simply _ignore_ it, as the behavioral implications of this option in the context of `bash` don't apply to PowerShell.

Steps to reproduce
------------------

```powershell
pwsh -l
pwsh --login
```

To see the problem in action with Visual Studio Code:

* Add `/usr/local/bin/pwsh` to `/etc/shells`
* Make PowerShell your default shell with `chsh -s /usr/local/bin/pwsh`
* Open a new terminal and close existing Visual Studio Code windows.
* Use `code <file>` to open a file for editing, where `<file>` must _not_ be a PowerShell file.
* Try to access the integrated terminal, which crashes due to invoking `pwsh` with `-l`.

Expected behavior
-----------------

An interactive PowerShell session should start, as if the option hadn't been specified.

Actual behavior
---------------

```none
Invalid argument '-l', did you mean:
...

The argument '--login' is not recognized as the name of a script file. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
...
```

Environment data
----------------

```powershell
PowerShell Core v6.1.0-rc.1 on macOS 10.13.6
PowerShell Core v6.1.0-rc.1 on Ubuntu 16.04.4 LTS
PowerShell Core v6.1.0-rc.1 on Microsoft Windows 10 Pro (64-bit; Version 1803, OS Build: 17134.165)
Windows PowerShell v5.1.17134.165 on Microsoft Windows 10 Pro (64-bit; Version 1803, OS Build: 17134.165)
```

**Comment by iSazonov** [comment]
If we add the `-l` whether users expect that we fully support PowerShell Core as login shell?

**Comment by mklement0** [comment]
We would have to document that it is an effective no-op.

De facto using `pwsh` as the default shell already works fine on macOS, after applying the workaround described in https://github.com/PowerShell/PowerShell/issues/6027#issuecomment-360638685

Eventually we'll need a documentation topic that describes all limitations with respect to using `pwsh` as the default shell on Unix-like platforms, currently summarized here: https://github.com/PowerShell/PowerShell/issues/975#issuecomment-331049792

**Comment by iSazonov** [comment]
@mklement0  Could you please clarify about 
> Technically, for PowerShell to detect being invoked as a login shell it must examine the equivalent of $0 (argv[0]) - the invocation name - to look for - as the 1st char.

Is it `pwsh -`?
Can this resolve the Issue too?

**Comment by iSazonov** [comment]
Seems we have already discussed that we could use different file names to address such scenarios (login shell, hide console windows, graphical shell and etc.). For the Issue we could create `pwshl` or `pwsh_login` runner.

**Comment by mklement0** [comment]
That would add unnecessary complexity:

* It requires creating, maintaining, shipping, documenting  a new executable, even though its sole purpose would be to ignore the `-l` / `--login` options.

* User will need to know and remember that they must specify that a different executable when changing their default shell to PowerShell with `chsh -s`, which is far from obvious (adding the right executable to `/etc/shells` as a prerequisite should be done by the installer, which may already be happening).

**Comment by iSazonov** [comment]
With -l option we only fix VS Code issue not Terminal.app 
I found the PowerShell Committee approvement https://github.com/PowerShell/PowerShell/issues/992#issuecomment-384455760

**Comment by mklement0** [comment]
_There is nothing to fix_ for Terminal.app or any other tools that use the `-`-prefix convention, because PowerShell doesn't and need not care about what `argv[0]` truly contains - it simply has no impact.

The approval you link to is in a way the _opposite_ of what we're discussing here: a dedicated `pwsh-np` would _suppress_ loading of profiles, whereas option `-l` in POSIX-like shells explicitly loads profiles (even in non-interactive shells).

**Comment by mklement0** [comment]
> These settings are not for end users.
> in this case, the end user does not need to know that another file is being launched.

That would be nice, but it's already not true for `pwsh-np`: As a script author, you must know to use _that_ as the executable in the shebang line in order to get predictable behavior that doesn't load `$PROFILE`.

Unfortunately, backward-compatibility concerns leaves us no choice there, so a separate executable is required.

Similarly, as stated, the end user would need to know that it's not `pwsh`, but `pwshl` (or whatever it would be) that must be passed to `chsh -s` _in order to make PowerShell the default shell._

Fortunately, in this case we _do_ have the option to support that with the existing executable, by implementing `-l` / `--login` as no-op switches.

**Comment by iSazonov** [comment]
Modern devices is locked - owners haven't root permissions. Corporate users too. Special devices (net devices, storage systems and so in) is locked too. Only a manufacturer configure the device. Or there is a config tools. Or installer. If user has a root permission one is high quality professional user. On windows systems we traditionally have GUI that helps eliminate errors in the configuration. In PowerShell we can add cmdlets to configure shells, schedule tasks and so on in right way.

Posix discribes Bash behaviors. If somebody want to implement something Posix-compatible one have to implement or port new-Bash. For any other shell like PowerShell we are forced to seek a compromise and implement only what is possible. 
If the team published PowerShell specifications under Posix, we would be so Posix-compatible :-) 
I think real way is to develop and publish RFC to standardize PowerShell shell behavior. Then any manufacturer of distributions and systems can embed PowerShell in a standard and compatible manner. It may even be quite different than Posix/Bash.
It should also be noted that in the 6.x version we must comply with backward compatibility, but we can do some useful workarounds (like you suggest). But this should be done so that in next 7.x version we could make more significant changes.

**Comment by mklement0** [comment]
@iSazonov:

Honestly, I have no idea what your first paragraph is talking about and how it relates to this discussion.

As for the second paragraph:

> Posix discribes Bash behaviors.

As an aside: No, POSIX comprises a _specification_ for a shell, whose executable is `/bin/sh`. Bash is a POSIX-_like_ shell, but even when you run it in POSIX compatibility mode it isn't fully POSIX-compliant.

> For any other shell like PowerShell we are forced to seek a compromise 

And implementing `-l` / `--login` as quiet no-op options is exactly such a _compromise_.

No one is looking for 100% POSIX compatibility, which fundamentally cannot be achieved by PowerShell.

What matters is that _in practice_ you can use it _as your default shell_ on _all of the supported platforms_, and the proposed change is a step in that direction.

If someone wants to take on the work of formalizing the extent to which PowerShell is POSIX-compatible, that's great, but doing so is not a prerequisite for the change proposed here.

**Comment by iSazonov** [comment]
@mklement0 Thanks for great discussion!

Yes, adding '-l' we resolve the Issue with VS Code.
My main fear was and is that with getting '-l' users will expect that we implementfully fully functional login shell. So first of all we need to create RFC and describe all shell modes. This will allow Community to use PowerShell in the standard way.

It is a question for PowerShell Committee.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee would accept adding `-l` that output a warning message akin to: "The -l switch was used to indicate starting a login shell, however, PowerShell Core 6 does not currently support this meaning that certain environment variables will not be available or complete.".  This would allow tools that expect `-l` to be there to work and pwsh to startup and allows us to support the complete work of a login shell in the future.

**Comment by mklement0** [comment]
Resolved in #9528
